WordPress/gutenberg
# Background images: add support for theme.json ref value resolution

## What?

Part of:

- https://github.com/WordPress/gutenberg/issues/54336

So, you know how you can "reference" another value in theme.json using `"ref"`?

Well,  this PR adds support for ref resolution to `"background"` style properties.


## Why?

The `"background"` style properties haven't supported this feature, until now!

The theme.json schema has already been updated to reflect this change.

## How?

By passing the resolved background image value to the style engine.

In the client (editor JS), I'm doing it long-hand, as you can see. It's a bit fugly. In both the frontend (PHP) and editor, the style engine does the work of escaping `background-image` values and also ensuring they're wrapped in `url()`. 

The `WP_Theme_JSON` class resolves values before passing to the style engine, however the editor does it the other way around. 

However, the style engine does not perform any "ref" resolution. And since the work of escaping and wrapping in `url()` needs to happen after ref resolution, we need to make an exception for background image.


## Alternatives

1. We add logic to `WP_Theme_JSON` and `use-global-styles-output.js` to sanitize/build `url()` values and bypass the style engine for `background-image` values. Involves duplicating code.

3. The style engine could handle ref resolution (and possibly defaults)  via a public API, used both internally and externally. Theme.json tree and defaults could be passed as options. 

Example (not prescriptive):

```js
// Input
const themeJson = {
   styles: {
    background: {
         backgroundImage: {
             url: 'http://some.host/someimage.gif',
         }
     }
   }
};

const blockStyles = {
    background: {
         backgroundImage: { ref: 'styles.background.backgroundImage' }
    }
};

// Output
const styles = getCSSRules( blockStyles, {
     themeJson,
} );

// Somewhere in getCSSRules

const resolvedRefValue = get( options.themeJson, blockStyles.background.backgroundImage.ref );
const backgroundImageValue = ``url( '${ encodeURI( safeDecodeURI( resolvedRefValue.url ) ) }' )`
```


## Testing Instructions

1. Create a theme.json with valid background ref values (see theme.json example below). The example below also includes a relative path to an image.
2. Ensure that the values resolve in the frontend and site editor (and post editor for the heck of it)
3. Check that, when you replace a block image in global styles, that the image appears correctly in the frontend and backend.

<details>

<summary>Example block HTML</summary>

```html


<div class="wp-block-group">

<p>Group</p>
</div>



<pre class="wp-block-verse">Verse</pre>



<blockquote class="wp-block-quote">
<p>Quote</p>
</blockquote>


```

</details>

<details>

<summary>Example theme.json</summary>

```json


{
	"$schema": "../../schemas/json/theme.json",
	"version": 3,
	"settings": {
		"appearanceTools": true,
		"layout": {
			"contentSize": "840px",
			"wideSize": "1100px"
		}
	},
	"styles": {

		"background": {
			"backgroundImage": {
				"url": "https://images.pexels.com/photos/105819/pexels-photo-105819.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr="
			},
			"backgroundSize": "cover",
			"backgroundAttachment": "fixed",
			"backgroundPosition": "top left"
		},
		"blocks": {
			"core/verse": {
				"background": {
					"backgroundImage": {
						"url": "file:./img/1024x768_emptytheme_test_image.jpg"
					},
					"backgroundAttachment": "fixed",
					"backgroundSize": "61px"
				},
				"color": {
					"text": "white"
				},
				"dimensions": {
					"minHeight": "100px"
				}
			},
			"core/quote": {
				"background": {
					"backgroundImage": {
						"ref": "styles.background.backgroundImage"
					}
				},
				"dimensions": {
					"minHeight": "100px"
				}
			},
			"core/group": {
				"background": {
					"backgroundImage": {
						"ref": "styles.blocks.core/verse.background.backgroundImage"
					},
					"backgroundAttachment": {
						"ref": "styles.blocks.core/verse.background.backgroundAttachment"
					},
					"backgroundSize": {
						"ref": "styles.blocks.core/verse.background.backgroundSize"
					},
					"backgroundPosition":  {
						"ref": "styles.background.backgroundPosition"
					}
				},
				"color": {
					"text": {
						"ref": "styles.blocks.core/verse.color.text"
					}
				},
				"dimensions": {
					"minHeight": "111px"
				}
			}
		}
	}
}





```

</details>

<img width="901" alt="Screenshot 2024-07-31 at 3 37 45 PM" src="https://github.com/user-attachments/assets/b1a9fa3a-9b5a-4b4c-ad56-8bbc6baf1402">

## TODO

Split this PR up into (in order of merge priority):

- [x] Abstracting the ref and theme URI resolution in the client and tests https://github.com/WordPress/gutenberg/pull/64182
- [x] Ensuring only user-added images get default block styles https://github.com/WordPress/gutenberg/pull/64192
- [x] Adding support for background theme.json ref resolution (this PR)
- [ ] Export `getCSSVarFromStyleValue` from style engine in place of `compileStyleValue` (new PR) https://github.com/WordPress/gutenberg/pull/64490
